readme.md
# accepts

[![tag](https://img.shields.io/github/tag/ako-deno/accepts.svg)](https://github.com/ako-deno/accepts/tags)
![accepts-ci](https://github.com/ako-deno/accepts/workflows/accepts-ci/badge.svg)
[![HitCount](http://hits.dwyl.com/ako-deno/accepts.svg)](http://hits.dwyl.com/ako-deno/accepts)

Higher level content negotiation for Deno using [negotiator](https://deno.land/x/negotiator). Based on `https://github.com/jshttp/accepts`.

In addition to negotiator, it allows:

- Allows type shorthands such as `json`.
- Returns `false` when no types match
- Treats non-existent headers as `*`

## API

```js
import { Accepts } from "https://deno.land/x/accepts/mod.ts";
```

### Accepts(headers: Headers)

Create a new `Accepts` object for the given `header`.

```js
const accept = new Accepts(header);
```

#### accept.charsets(charsets?: string[]): string[] | string | boolean

Return the first accepted charset. If nothing in `charsets` is accepted,
then `false` is returned.

Return the charsets that the request accepts, in the order of the client's
preference (most preferred first).

#### accept.encodings(encodings?: string[]): string[] | string | boolean

Return the first accepted encoding. If nothing in `encodings` is accepted,
then `false` is returned.

Return the encodings that the request accepts, in the order of the client's
preference (most preferred first).

#### accept.languages(languages?: string[]): string[] | string | boolean

Return the first accepted language. If nothing in `languages` is accepted,
then `false` is returned.

Return the languages that the request accepts, in the order of the client's
preference (most preferred first).

#### accept.types(types?: string[]): string[] | string | boolean

Return the first accepted type (and it is returned as the same text as what
appears in the `types` array). If nothing in `types` is accepted, then `false`
is returned.

The `types` array can contain full MIME types or file extensions. Any value
that is not a full MIME types is passed to [mime_types](https://deno.land/x/media_types)'s.lookup`.

Return the types that the request accepts, in the order of the client's
preference (most preferred first).

## Examples

### Simple type negotiation

This simple example shows how to use `accepts` to return a different typed
respond body based on what the client wants to accept. The server lists it's
preferences in order and will get back the best match between the client and
server.

```js
import {
  serve,
  Response,
} from "https://deno.land/std@0.69.0/http/server.ts";
import { Accepts } from "../mod.ts";

const server = serve("127.0.0.1:3000");
console.log("Server listening on: 3000");

for await (const req of server) {
  const accept = new Accepts(req.headers);
  const res: Response = {
    headers: new Headers(),
  };
  let type = accept.types(["json", "html"]);
  if (type) {
    type = typeof type === "string" ? type : type[0];
    switch (type) {
      case "json":
        res.body = '{"hello":"world!"}';
        res.headers!.set("Content-Type", "application/json");
        break;
      case "html":
        res.body = "<b>hello, world!</b>";
        res.headers!.set("Content-Type", "text/html");
        break;
    }
  } else {
    // the fallback is text/plain, so no need to specify it above
    res.body = "hello, world!";
    res.headers!.set("Content-Type", "text/plain");
  }
  req.respond(res).catch(() => {});
}
```

You can test this out with the cURL program:
```sh
curl -I -H'Accept: text/html' http://localhost:3000/
```

# License

[MIT](./LICENSE)


History.md

2.1.1 / 2021-08-08
==================

  * bump media_types dep (#4)

2.1.0 / 2020-09-17
==================

  * (feat): upgrade imports to support Deno 1.4.0 strict type imports (#3)

2.0.0 / 2020-07-17
==================

  * doc: update doc and example on the return type
  * feat: only return false insted of boolean
  * breaking: make return value type compatible with original accepts

1.0.1 / 2020-07-14
==================

  * fixed version test deps
  * add hits
  * Merge pull request #1 from ako-deno/ngot-patch-1
  * doc: use deno.land/x url

1.0.0 / 2020-05-23
==================

  * doc: add for Deno
  * doc: fix example link
  * foc: fix example and api doc
  * refactor: remove alias and false to []
  * init


test/deps.ts
// std
export {
  assertEquals,
} from "https://deno.land/std@0.69.0/testing/asserts.ts";


test/type_test.ts
import {
  assertEquals,
} from "./deps.ts";
import { Accepts } from "../mod.ts";

const { test } = Deno;

test("accepts.types() with no arguments when Accept is populated should return all accepted types", function () {
  const header = createRequest(
    "application/*;q=0.2, image/jpeg;q=0.8, text/html, text/plain",
  );
  const accept = new Accepts(header);
  assertEquals(
    accept.types(),
    ["text/html", "text/plain", "image/jpeg", "application/*"],
  );
});

test("accepts.types() with no arguments when Accept not in request should return */*", function () {
  const header = createRequest();
  const accept = new Accepts(header);
  assertEquals(accept.types(), ["*/*"]);
});

test("accepts.types() with no arguments when Accept is empty should returnfalse", function () {
  const header = createRequest("");
  const accept = new Accepts(header);
  assertEquals(accept.types(), []);
});

test("accepts.types() with no valid types when Accept is populated should returnfalse", function () {
  const header = createRequest(
    "application/*;q=0.2, image/jpeg;q=0.8, text/html, text/plain",
  );
  const accept = new Accepts(header);
  assertEquals(accept.types(["image/png", "image/tiff"]), false);
});

test("accepts.types() with no valid types when Accept is not populated should return the first type", function () {
  const header = createRequest();
  const accept = new Accepts(header);
  assertEquals(
    accept.types(["text/html", "text/plain", "image/jpeg", "application/*"]),
    "text/html",
  );
});

test("accepts.types() when extensions are given should convert to mime types", function () {
  const header = createRequest("text/plain, text/html");
  const accept = new Accepts(header);
  assertEquals(accept.types(["html"]), "html");
  assertEquals(accept.types([".html"]), ".html");
  assertEquals(accept.types(["txt"]), "txt");
  assertEquals(accept.types([".txt"]), ".txt");
  assertEquals(accept.types(["png"]), false);
  assertEquals(accept.types(["bogus"]), false);
});

test("accepts.types() when an array is given should return the first match", function () {
  const header = createRequest("text/plain, text/html");
  const accept = new Accepts(header);
  assertEquals(accept.types(["png", "text", "html"]), "text");
  assertEquals(accept.types(["png", "html"]), "html");
  assertEquals(accept.types(["bogus", "html"]), "html");
});

test("accepts.types() when present in Accept as an exact match should return the type", function () {
  const header = createRequest("text/plain, text/html");
  const accept = new Accepts(header);
  assertEquals(accept.types(["text/html"]), "text/html");
  assertEquals(accept.types(["text/plain"]), "text/plain");
});

test("accepts.types() when present in Accept as a type match should return the type", function () {
  const header = createRequest("application/json, */*");
  const accept = new Accepts(header);
  assertEquals(accept.types(["text/html"]), "text/html");
  assertEquals(accept.types(["text/plain"]), "text/plain");
  assertEquals(accept.types(["image/png"]), "image/png");
});

test("accepts.types() when present in Accept as a subtype match should return the type", function () {
  const header = createRequest("application/json, text/*");
  const accept = new Accepts(header);
  assertEquals(accept.types(["text/html"]), "text/html");
  assertEquals(accept.types(["text/plain"]), "text/plain");
  assertEquals(accept.types(["image/png"]), false);
});

function createRequest(type?: string) {
  const header = new Headers();
  if (type !== undefined) {
    header.set("accept", type);
  }
  return header;
}


test/charset_test.ts
import {
  assertEquals,
} from "./deps.ts";
import { Accepts } from "../mod.ts";

const { test } = Deno;

test("accepts.charsets() with no arguments when Accept-Charset is populated should return accepted types", function () {
  const header = createRequest("utf-8, iso-8859-1;q=0.2, utf-7;q=0.5");
  const accept = new Accepts(header);
  assertEquals(accept.charsets(), ["utf-8", "utf-7", "iso-8859-1"]);
});

test("accepts.charsets() when Accept-Charset is not in request should return *", function () {
  const header = createRequest();
  const accept = new Accepts(header);
  assertEquals(accept.charsets(), ["*"]);
});

test("accepts.charsets() when Accept-Charset is empty should return an empty array", function () {
  const header = createRequest("");
  const accept = new Accepts(header);
  assertEquals(accept.charsets(), []);
});

test("accepts.charsets() with multiple arguments when Accept-Charset is populated if any types match should return the best fit", function () {
  const header = createRequest("utf-8, iso-8859-1;q=0.2, utf-7;q=0.5");
  const accept = new Accepts(header);
  assertEquals(accept.charsets(["utf-7", "utf-8"]), "utf-8");
});

test("accepts.charsets() with multiple arguments when Accept-Charset is populated if no types match should return []", function () {
  const header = createRequest("utf-8, iso-8859-1;q=0.2, utf-7;q=0.5");
  const accept = new Accepts(header);
  assertEquals(accept.charsets(["utf-16"]), false);
});

test("accepts.charsets() with multiple arguments when Accept-Charset is populated when Accept-Charset is not populated should return the first type", function () {
  const header = createRequest();
  const accept = new Accepts(header);
  assertEquals(accept.charsets(["utf-7", "utf-8"]), "utf-7");
});

test("accepts.charsets() with an array should return the best fit", function () {
  const header = createRequest("utf-8, iso-8859-1;q=0.2, utf-7;q=0.5");
  const accept = new Accepts(header);
  assertEquals(accept.charsets(["utf-7", "utf-8"]), "utf-8");
});

function createRequest(charset?: string) {
  const header = new Headers();
  if (charset !== undefined) {
    header.set("accept-charset", charset);
  }
  return header;
}


example/server.ts
import {
  serve,
  Response,
} from "https://deno.land/std@0.69.0/http/server.ts";
import { Accepts } from "../mod.ts";

const server = serve("127.0.0.1:3000");
console.log("Server listening on: 3000");

for await (const req of server) {
  const accept = new Accepts(req.headers);
  const res: Response = {
    headers: new Headers(),
  };
  let type = accept.types(["json", "html"]);
  if (type) {
    type = typeof type === "string" ? type : type[0];
    switch (type) {
      case "json":
        res.body = '{"hello":"world!"}';
        res.headers!.set("Content-Type", "application/json");
        break;
      case "html":
        res.body = "<b>hello, world!</b>";
        res.headers!.set("Content-Type", "text/html");
        break;
    }
  } else {
    // the fallback is text/plain, so no need to specify it above
    res.body = "hello, world!";
    res.headers!.set("Content-Type", "text/plain");
  }
  req.respond(res).catch(() => {});
}


mod.ts
/*!
 * Based on https://github.com/jshttp/accepts/blob/master/index.js
 * Copyright(c) 2014 Jonathan Ong
 * Copyright(c) 2015 Douglas Christopher Wilson
 * Copyright(c) 2020 Henry Zhuang
 * MIT Licensed
 */

import { Negotiator, lookup } from "./deps.ts";

/**
 * Create a new Accepts object for the given headers.
 *
 * @param {Headers} headers
 * @public
 */
export class Accepts {
  private headers: Headers;
  private negotiator: Negotiator;

  constructor(headers: Headers) {
    this.headers = headers;
    this.negotiator = new Negotiator(headers);
  }

  /**
   * Check if the given `type(s)` is acceptable, returning
   * the best match when true, otherwise `undefined`, in which
   * case you should respond with 406 "Not Acceptable".
   *
   * The `type` value may be a single mime type string
   * such as "application/json", the extension name
   * such as "json" or an array `["json", "html", "text/plain"]`. When a list
   * or array is given the _best_ match, if any is returned.
   *
   * Examples:
   *
   *     // Accept: text/html
   *     this.types(['html']);
   *     // => ["html"]
   *
   *     // Accept: text/*, application/json
   *     this.types(['html']);
   *     // => ["html"]
   *     this.types(['text/html']);
   *     // => ["text/html"]
   *     this.types(['json', 'text']);
   *     // => ["json"]
   *     this.types(['application/json']);
   *     // => ["application/json"]
   *
   *     // Accept: text/*, application/json
   *     this.types(['image/png']);
   *     this.types(['png']);
   *     // => []
   *
   *     // Accept: text/*;q=.5, application/json
   *     this.types(['html', 'json']);
   *     // => ["json"]
   *
   * @param {Array} types...
   * @return {Array|String|Boolean}
   * @public
   */
  types(types?: string[]): string[] | string | boolean {
    // no types, return all requested types
    if (!types || types.length === 0) {
      return this.negotiator.mediaTypes();
    }

    // no accept header, return first given type
    if (!this.headers.get("accept")) {
      return types[0];
    }

    const mimes = types
      .map(extToMime);
    const accepts = this.negotiator.mediaTypes(
      (mimes.filter((t) => t && validMime(t))) as string[],
    );
    const first = accepts[0];
    return first ? types[mimes.indexOf(first)] : false;
  }

  /**
   * Return accepted encodings or best fit based on `encodings`.
   *
   * Given `Accept-Encoding: gzip, deflate`
   * an array sorted by quality is returned:
   *
   *     ['gzip', 'deflate']
   *
   * @param {Array} encodings...
   * @return {Array|String|Boolean}
   * @public
   */
  encodings(encodings?: string[]): string[] | string | boolean {
    // no encodings, return all requested encodings
    if (!encodings || encodings.length === 0) {
      return this.negotiator.encodings();
    }
    return this.negotiator.encodings(encodings)[0] || false;
  }

  /**
   * Return accepted charsets or best fit based on `charsets`.
   *
   * Given `Accept-Charset: utf-8, iso-8859-1;q=0.2, utf-7;q=0.5`
   * an array sorted by quality is returned:
   *
   *     ['utf-8', 'utf-7', 'iso-8859-1']
   *
   * @param {Array} charsets...
   * @return {Array|String|Boolean}
   * @public
   */
  charsets(charsets?: string[]): string[] | string | boolean {
    // no charsets, return all requested charsets
    if (!charsets || charsets.length === 0) {
      return this.negotiator.charsets();
    }
    return this.negotiator.charsets(charsets)[0] || false;
  }

  /**
   * Return accepted languages or best fit based on `langs`.
   *
   * Given `Accept-Language: en;q=0.8, es, pt`
   * an array sorted by quality is returned:
   *
   *     ['es', 'pt', 'en']
   *
   * @param {Array} langs...
   * @return {Array|String|Boolean}
   * @public
   */
  languages(languages?: string[]): string[] | string | boolean {
    // no languages, return all requested languages
    if (!languages || languages.length === 0) {
      return this.negotiator.languages();
    }

    return this.negotiator.languages(languages)[0] || false;
  }
}

/**
 * Convert extnames to mime.
 *
 * @param {String} type
 * @return {String|undefined}
 * @private
 */

function extToMime(type: string): string | undefined {
  return type.indexOf("/") === -1 ? lookup(type) : type;
}

/**
 * Check if mime is valid.
 *
 * @param {String} type
 * @return {String}
 * @private
 */

function validMime(type?: string): boolean {
  return typeof type === "string";
}
